gobblin-core/src/main/java/gobblin/writer/AvroHdfsDatePartitionedWriter.java

```diff
@@ -0,0 +1,271 @@
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.hadoop.fs.Path;
+
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Maps;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.configuration.State;
+import gobblin.util.AvroUtils;
+import gobblin.util.ForkOperatorUtils;
+
+
+/**
+ * Implementation of {@link DataWriter} that writes data into a date-partitioned directory structure based on the value
+ * of a specific field in each Avro record.
+ *
+ * <p>
+ *
+ * The writer takes in each Avro record, extracts out a specific field, and based on the the value of the column it
+ * writes the data into a specific file. The column is assumed to be of type {@link Long}, and the value is treated as
+ * a timestamp. By default, the timestamp is converted into a directory of the form [yyyy]/[MM][dd]/. However, the
+ * actual pattern used to construct the path is specified by {@link ConfigurationKeys#WRITER_PARTITION_PATTERN}. This
+ * class uses Joda's {@link DateTimeFormatter} to parse a timestamp, and convert it to a {@link String}. The
+ * {@link String} is then converted to a Path, which specifies where the record will be written.
+ *
+ * <p>
+ *
+ * By default, The complete output path for a directory is "/baseFilePath/daily/[yyyy]/[MM]/[dd]/fileName.avro". The
+ * baseFilePath is specified by the configuration key {@link ConfigurationKeys#WRITER_FILE_PATH}. The writer uses the
+ * configuration key {@link ConfigurationKeys#WRITER_PARTITION_COLUMN_NAME} to determine the name of the column to
+ * partition by. The "daily" portion of the path is also configurable, and can be changed by setting
+ * {@link ConfigurationKeys#DEFAULT_WRITER_PARTITION_LEVEL}.
+ *
+ * <p>
+ *
+ * The class works by maintaining a {@link Map} of {@link Path}s to {@link DataWriter}s called {@link #pathToWriterMap}.
+ * The {@link AvroHdfsDatePartitionedWriter#write(GenericRecord)} takes each record, extracts out the the value of the
+ * field specified by {@link ConfigurationKeys#WRITER_PARTITION_COLUMN_NAME}, and constructs a {@link Path} based on the
+ * field value. If a {@link DataWriter} already exists for the path, the record is simply written to the existing
+ * {@link DataWriter}. If one does not exist, then a new {@link DataWriter} is created, added to the Map, and then the
+ * input record is written to the writer.
+ *
+ * <p>
+ *
+ * By using the above approach methods such as {@link DataWriter#commit()}, {@link DataWriter#close()}, etc. are simple.
+ * The implementation is to simply iterate over the values of {@link #pathToWriterMap} and call the corresponding method
+ * on each {@link DataWriter}.
+ */
+public class AvroHdfsDatePartitionedWriter implements DataWriter<GenericRecord> {
```

Comment:
A better name is `AvroHdfsTimePartitionedWriter`, since the partition is not necessarily by date.